bentoml/_internal/frameworks/onnxmlir.py

```diff
@@ -0,0 +1,396 @@
+
+# TODO: mount session to docker container and use prebuilt?
+DEV_MLIR_IMAGE = "onnxmlirczar/onnx-mlir-dev:latest"
+
+# release docker images to compile the binary
+RELEASE_MLIR_IMAGE = "onnxmlirczar/onnx-mlir:latest"
+DOCKER_WORKDIR = "/workdir"
+DOCKER_OUTDIR = "/output"
+COMPILE_ARG = ["--O3", "--EmitLib"]
+
+# onnx-mlir constant
+if sys.platform == "win32":
+    _onnx_mlir_binary = "onnx-mlir.exe"
+else:
+    _onnx_mlir_binary = "onnx-mlir"
+
+# There are loosely a few defined environment variables
+# from onnx/onnx-mlir that are available throughout its codebase.
+# here are for documentation
+ACCEPTED_ONNX_ENVVAR = ["ONNX_MLIR_HOME", "ONNX_MLIR_BIN"]
+
+
+@attr.define
+class ONNXMLirOptions(ModelOptions):
+    use_default_entry_point: bool = attr.field(default=True)
+    use_docker: bool = attr.field(default=False)
+
+
+def check_environment() -> tuple[str | None, str | None]:
+    _runtime_spec = find_spec("PyRuntime")
+    if _runtime_spec and _runtime_spec.origin:
+        # best effort to find onnx-mlir binary
+        logger.debug(
+            "BentoML will provide best effort to find onnx-mlir runtime. This behaviour depends on the output 'build' directory of onnx-mlir are kept as default. If you customized the output directory, you need to set either ONNX_MLIR_HOME or ONNX_MLIR_BIN manually."
+        )
+        pyruntime_binary = os.path.realpath(_runtime_spec.origin)
+        _onnx_mlir_runtime = os.path.dirname(pyruntime_binary)
+        _onnx_mlir = os.path.join(
+            os.path.dirname(_onnx_mlir_runtime), "bin", _onnx_mlir_binary
+        )
+    elif "ONNX_MLIR_HOME" in os.environ:
+        _onnx_mlir_runtime = os.path.join(os.environ["ONNX_MLIR_HOME"], "lib")
+        _onnx_mlir = os.path.join(
+            os.environ["ONNX_MLIR_HOME"], "bin", _onnx_mlir_binary
+        )
+    elif "ONNX_MLIR_BIN" in os.environ:
+        _onnx_mlir_runtime = os.path.join(
+            os.path.dirname(os.environ["ONNX_MLIR_BIN"]), "lib"
+        )
+        _onnx_mlir = os.path.join(os.environ["ONNX_MLIR_BIN"], _onnx_mlir_binary)
+    else:
+        return None, None
+    return _onnx_mlir, _onnx_mlir_runtime
+
+
+warning = """\
+The compiled Python runtime 'PyRuntime' is required in order to use module `bentoml.onnxmlir`. Please set either ONNX_MLIR_HOME or ONNX_MLIR_BIN so that BentoML can find onnx-mlir binary and runtime library correctly:
+
+1. Set ONNX_MLIR_HOME to the path of 'build' directory of onnx-mlir. The 'build' directory refers to parent folder containing 'bin', 'lib' in which ONNX-MLIR executables and libraries can be found, typically '/path/to/onnx-mlir/build/Debug'.
+2. Set ONNX_MLIR_BIN to 'onnx-mlir/build/Debug/bin' directory. After, add $ONNX_MLIR_BIN to PATH environment variable: "export PATH=$ONNX_MLIR_BIN:$PATH" and reload current shell.
+"""
+
+ONNX_MLIR, ONNX_MLIR_RUNTIME = check_environment()
+
+
+@contextmanager
+def setup_execution_session() -> t.Generator[t.Type[ExecutionSession], None, None]:
```

Comment:
Since we are creating a context manager here, I feel that we can definitely mount the docker process and use the prebuilt pyruntime. Note that this will only work for python 3.8 since the compiled PyRuntime on the docker image is for python 3.8.

seems it didn't work with other python version other than 3.8.

Reply:
As observed PyRuntime has a hard compatibility requirement on the python level it was built for. We may be able build the onnx-mlir with multiple levels to help testing in the future.

Reply:
what do you mean by this? Is it possible to support multiple python version upstream?

It seems to me that if bentoml provided a docker images that contains onnxmlir binary for all of the support python this would be the best.

Reply:
but again this is still explorable components I'm hacking around.

Reply:
I think the onnx-mlir image would have to include the .so pyruntime needs for each desired python level. Let me take a look at that and see if there is something that can be explored

Reply:
sg.